testing/src/KokkosFFT_CountErrors.hpp

```diff
@@ -0,0 +1,514 @@
+  using a_value_type = typename AViewType::non_const_value_type;
+  using b_value_type = typename BViewType::non_const_value_type;
+
+  using AErrorViewType = Kokkos::View<a_value_type*, ExecutionSpace>;
+  using BErrorViewType = Kokkos::View<b_value_type*, ExecutionSpace>;
+  using CountViewType  = Kokkos::View<std::size_t**, ExecutionSpace>;
+  using CountType      = Kokkos::View<std::size_t, ExecutionSpace>;
+
+  AViewType m_a;
+  BViewType m_b;
+
+  AErrorViewType m_a_error;
+  BErrorViewType m_b_error;
+  CountViewType m_loc_error;
+  CountType m_count;
+
+  double m_rtol;
+  double m_atol;
+
+  using policy_type =
+      Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<iType>>;
+
+  /// \brief Constructs a FindErrors object and performs the error detection.
+  /// This constructor initializes the output error views and launches a
+  /// parallel kernel to scan through the input view elements. For each element,
+  /// if the difference between the two views exceeds the specified tolerance,
+  /// the corresponding error values and their index are recorded.
+  ///
+  /// \param a [in] The first Kokkos view containing data.
+  /// \param b [in] The second Kokkos view containing data to compare against.
+  /// \param nb_errors [in] The maximum number of errors expected.
+  /// \param rtol [in] Relative tolerance for the comparison (default 1.e-5).
+  /// \param atol [in] Absolute tolerance for the comparison (default 1.e-8).
+  /// \param space [in] The execution space used to launch the parallel kernel.
+  FindErrors(const AViewType& a, const BViewType& b, const iType nb_errors,
+             double rtol = 1.e-5, double atol = 1.e-8,
+             const ExecutionSpace space = ExecutionSpace())
+      : m_a(a),
+        m_b(b),
+        m_a_error("a_error", nb_errors),
+        m_b_error("b_error", nb_errors),
+        m_loc_error("loc_error", nb_errors, 2),
+        m_count("count"),
+        m_rtol(rtol),
+        m_atol(atol) {
+    Kokkos::parallel_for("FindErrors-1D", policy_type(space, 0, m_a.extent(0)),
+                         *this);
+  }
+
+  ///\brief Executes the element-wise comparison for the given index.
+  /// This operator is invoked in parallel by Kokkos. For each index, it
+  /// compares the corresponding elements from the two views. If the absolute
+  /// difference exceeds the tolerance, it stores the error values and their
+  /// index.
+  ///
+  ///\param i0 [in] The index of the element in the views.
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const iType i0) const {
+    auto tmp_a     = m_a(i0);
+    auto tmp_b     = m_b(i0);
+    bool not_close = are_not_close(tmp_a, tmp_b, m_rtol, m_atol);
+    if (not_close) {
+      std::size_t count     = Kokkos::atomic_load(m_count.data());
+      m_a_error(count)      = tmp_a;
+      m_b_error(count)      = tmp_b;
+      m_loc_error(count, 0) = i0;
+      m_loc_error(count, 1) = i0;
+      Kokkos::atomic_fetch_add(m_count.data(), 1);
```

Comment:
If you just load `count`, multiple threads can end up with the same `count` and write to the same locations
```suggestion
      std::size_t count     = Kokkos::atomic_fetch_add(m_count.data(), 1);
      m_a_error(count)      = tmp_a;
      m_b_error(count)      = tmp_b;
      m_loc_error(count, 0) = i0;
      m_loc_error(count, 1) = i0;
```

Reply:
That is true. I will fix accordingly.